### PR TITLE
dynamic import syntax transformation in Ember CLI

### PR DESCRIPTION
**Pull request by frederikbosch**
This RFC proposes implementing a dynamic import syntax transformation in Ember CLI.

### CODE REVIEW

**Comment by locks** [comment]
@frederikbosch the structure of this document does not match the current template, did you make sure to update your repository before branching?

**Comment by frederikbosch** [comment]
@locks I took one of the earlier RFCs as template. I guess I missed something...

**Comment by frederikbosch** [comment]
@locks Maybe you can also tell if you like the general idea of this RFC. I love to update, and spend more time on it, if people feel this is heading into the right direction.

**Comment by stefanpenner** [comment]
The general idea is good.

Some Questions:

* what does the on-disk format look like for "lazy loaded modules", different folder or folders?
* how do we teach this
* provide some concrete motivating examples (loading some large `wysiwyg` editor code only in some places)
* how does this related/work with/or contradict https://github.com/emberjs/rfcs/pull/158

**Comment by frederikbosch** [comment]
@stefanpenner I updated the RFC with one concrete example and a list of other possible directions. Since Ember is already using import/export syntax, this RFC will not have a high impact. It only proposes to allow _dynamic_ imports, which is an upcoming javascript feature. Therefore, this should not contradict any other work, rfc or feature. Furthermore, there is no additional work required to teach this. Those who will need to import some module dynamically, they can be redirected to the standard.

**Comment by frederikbosch** [comment]
Extended the number of examples even further and also added a future section

File: active/0000-dynamic-import-syntaxt.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+```js
+import Ember from 'ember';
+
+export default Ember.Service.extend({
+  someMethodOnThisService() {
+    return Ember.RSVP.Promise((resolve, reject) => {
+      fetch('/resource').then((data) => {
+        if (data.variable === 1) {
+          import('app/services/x').then(service => resolve(service.default), reject);
```
Comment (kratiahuja) [inline active/0000-dynamic-import-syntaxt.md:43]: How is this different than doing something like this in an app:

```
   import require from 'require';
   ....
   if (data.variable === 1) {
     require('app/services/x').then
     ...
  }
```

What's the motivation to introduce a new public API when one already exists (Ember's loader API)?

**Comment by frederikbosch** [comment]
@kratiahuja Thanks for your feedback. What's the motivation to use import/export syntax instead of define and require (Ember's loader API) everywhere? That is that Ember is complying to newest JS syntax. Same is true for dynamic imports. It's an upcoming JS feature.

I think eventually Ember's loader API will only exist as polyfill for import/export. Or, when [the official loader API will land](https://whatwg.github.io/loader/) as implementation for that API.

**Comment by frederikbosch** [comment]
With the [recent work by @rwjblue](https://github.com/ember-cli/ember-cli/pull/6828), this RFC might become even more relevant. I would love to create a PR for this RFC if it gets accepted.

**Comment by rwjblue** [comment]
I'm in favor of this, and am happy to push things forward at the next ember-cli core team meeting.

@frederikbosch - Can you please prepare a PR to ember-cli-babel (current master branch) implementing this? Its often quite useful to see an implementation in the wild, to know that an RFC has been properly informed by the implementation details...

**Comment by Pauan** [comment]
Hello. I work for [amCharts](https://www.amcharts.com/). We have recently published a pre-release of the next big version of our software library (version 4).

Our version 4 library uses the dynamic `import` syntax. This works perfectly fine in Webpack, but one of our customers is using Ember, and it does not work for them.

We believe this sort of situation will happen more frequently in the future, as more libraries start using the dynamic `import` syntax.

**Comment by Pauan** [comment]
P.S. We use the dynamic `import` to load npm modules, like this: `import("fabric")`, so if this RFC is accepted Ember should make sure that it also supports dynamic `import` of npm modules.

**Comment by frederikbosch** [comment]
@Pauan The simple (untested) Ember solution to that is the following.

```js
// vendor/dynamic-import.js
window['import'] = function(moduleName) {
  return Ember.RSVP.Promise.resolve(window.require(moduleName));
}

// ember-cli-build.js
app.import('vendor/dynamic-import.js');

**Comment by Pauan** [comment]
@frederikbosch I don't think that will work, because `import("fabric")` is not valid syntax in ES5, so it cannot be polyfilled like that. The transformation *must* happen at build-time.

**Comment by frederikbosch** [comment]
Why not? Dynamic import is a regular function, and the above is a polyfill for it. If `fabric` is added as an Ember plugin or added as vendor in `ember-cli-build.js`, it *will* be transpiled during builds and you should be able to get this to work.

**Comment by Pauan** [comment]
@frederikbosch Dynamic `import` is *not* a regular function. It is special magical syntax, just like top-level `import` and `export`. To be more specific, [`import` is a keyword](https://www.ecma-international.org/ecma-262/6.0/#sec-keywords), and thus it *cannot* be used as a variable or function name.

**Comment by frederikbosch** [comment]
@Pauan You're right, my solution was too simple. I thought that worked, and I still think it did until browsers like Chrome started implementing the feature.

**Comment by Pauan** [comment]
To add a bit more clarification: yes you can use `window["import"] = ...` to *define* an `import` function, however when you *call* `import("foo")` one of two things will happen:

1. In browsers that *don't* support dynamic `import` you will get a SyntaxError, because `import` is a keyword.

2. In browsers that *do* support dynamic `import` you will get a "module could not be found" error.

In either case the `window["import"]` function will *never* be called. Thus, it is impossible to create a polyfill for dynamic `import`.

**Comment by frederikbosch** [comment]
You might also consider using parameters of [ember-cli-babel](https://github.com/babel/ember-cli-babel).

**Comment by frederikbosch** [comment]
And then use the [babel plugin](https://www.npmjs.com/package/babel-plugin-dynamic-import-amd) that I already created.

**Comment by Pauan** [comment]
@frederikbosch Yeah, that might work. I wasn't aware that an AMD version of `babel-plugin-dynamic-import` existed, thanks!

In that case, shouldn't it be very easy for Ember to support dynamic `import`? It can just enable the `babel-plugin-dynamic-import-amd` plugin by default.

**Comment by frederikbosch** [comment]
Yes it is. But it had to go through RFC. And I created this RFC, including with the plugin that should provide the functionality. But then I should also create a PR to `ember-cli-babel`, including tests, and that is I believe what I have not yet.

**Comment by rwjblue** [comment]
We are working on closing the ember-cli/rfcs repo in favor of using a single central RFC's repo for everything. This was laid out in https://emberjs.github.io/rfcs/0300-rfc-process-update.html. 

Sorry for the troubles, but would you mind reviewing to see if this is still something we need, and if so migrating this over to emberjs/rfcs?

Thank you!

**Pull request closed without merging**